vm/systemSmartContracts/auction.go

```diff
@@ -1165,7 +1212,158 @@ func (s *stakingAuctionSC) claim(args *vmcommon.ContractCallInput) vmcommon.Retu
 	return vmcommon.Ok
 }
 
-func (s *stakingAuctionSC) calculateNodePrice(bids []AuctionData) (*big.Int, error) {
+func (s *stakingAuctionSC) unStakeTokens(args *vmcommon.ContractCallInput) vmcommon.ReturnCode {
+	if !s.flagStakingV2.IsSet() {
+		//unStakeTokens function will become available after enabling staking v2
+		s.eei.AddReturnMessage("invalid method to call")
+		return vmcommon.UserError
+	}
+
+	if len(args.Arguments) != 1 {
+		s.eei.AddReturnMessage("should have specified one argument containing the unstake value")
+		return vmcommon.UserError
+	}
+	if args.CallValue.Cmp(zero) != 0 {
+		s.eei.AddReturnMessage(vm.TransactionValueMustBeZero)
+		return vmcommon.UserError
+	}
+	registrationData, err := s.getOrCreateRegistrationData(args.CallerAddr)
+	if err != nil {
+		s.eei.AddReturnMessage("cannot get registration data: error " + err.Error())
+		return vmcommon.UserError
+	}
+	if len(registrationData.RewardAddress) == 0 {
+		s.eei.AddReturnMessage("key is not registered, unStakeTokens is not possible")
+		return vmcommon.UserError
+	}
+	err = s.eei.UseGas(s.gasCost.MetaChainSystemSCsCost.Claim)
+	if err != nil {
+		s.eei.AddReturnMessage(vm.InsufficientGasLimit)
+		return vmcommon.OutOfGas
+	}
+	if registrationData.TotalUnstaked == nil {
+		registrationData.TotalUnstaked = big.NewInt(0)
+	}
+
+	maxValUnstake := s.computeMaxValueThatCanBeUnStaked(registrationData)
+	unStakeValue := big.NewInt(0).SetBytes(args.Arguments[0])
+	//unstakeValueIsOk will be true if unstake value is larger than the one provided in the config OR is exactly the
+	// value that was left to be unstaked
+	unstakeValueIsOk := unStakeValue.Cmp(s.minUnstakeTokensValue) >= 0 || unStakeValue.Cmp(maxValUnstake) == 0
+	if !unstakeValueIsOk {
+		s.eei.AddReturnMessage("can not unstake the provided value either because is under the minimum threshold or " +
+			"is not the value left to be unStaked")
+		return vmcommon.UserError
+	}
+
+	registrationData.TotalStakeValue.Sub(registrationData.TotalStakeValue, unStakeValue)
+	if registrationData.TotalStakeValue.Cmp(zero) < 0 {
+		s.eei.AddReturnMessage("contract error on unStakeTokens function, total stake < 0")
+		return vmcommon.UserError
+	}
+
+	registrationData.TotalUnstaked.Add(registrationData.TotalUnstaked, unStakeValue)
+	registrationData.UnstakedInfo = append(
+		registrationData.UnstakedInfo,
+		&UnstakedValue{
+			UnstakedNonce: s.eei.BlockChainHook().CurrentNonce(),
+			UnstakedValue: unStakeValue,
+		},
+	)
+
+	err = s.saveRegistrationData(args.CallerAddr, registrationData)
+	if err != nil {
+		s.eei.AddReturnMessage("cannot save registration data: error " + err.Error())
+		return vmcommon.UserError
+	}
+
+	return vmcommon.Ok
+}
+
+func (s *stakingAuctionSC) computeMaxValueThatCanBeUnStaked(registrationData *AuctionDataV2) *big.Int {
+	maxVal := big.NewInt(0).Sub(registrationData.TotalStakeValue, registrationData.LockedStake)
+	maxVal.Sub(maxVal, registrationData.TotalUnstaked)
```

Comment:
I think this substract is not needed as the value of TotalStakeValue it is already adjusted in line 1259 and then saved in line 1274, after each call of method unStakeTokens

Reply:
it is required because I need the maximum value that can be unstaked. this instruction
```GO
unstakeValueIsOk := unStakeValue.Cmp(s.minUnstakeTokensValue) >= 0 || unStakeValue.Cmp(maxValUnstake) == 0
```
evaluates if the unstake value is at least the minimum unstake value allowed or is exactly the remaining unstakeable value as to allow collecting the "dust" left. Indeed, I need to add an extra check as to not let the user call unstake with a larger value that can actually be unstake. Thanks for pointing this out! :+1: